src/object-store/src/util.rs

```diff
@@ -44,8 +34,42 @@ pub fn normalize_dir(dir: &str) -> String {
 pub fn join_dir(parent: &str, child: &str) -> String {
     // Always adds a `/` to the output path.
     let output = format!("{parent}/{child}/");
-    // We call opendal's normalize_root which keep the last `/`.
-    opendal::raw::normalize_root(&output)
+    normalize_dir(&output)
+}
+
+/// Modified from the `opendal::raw::normalize_root`
+///
+/// # The different
+///
+/// It doesn't always append `/` ahead of the path,
+/// It only keeps `/` ahead if the original path starts with `/`.
+///
+/// Make sure root is normalized to style like `/abc/def/`.
+///
+/// # Normalize Rules
+///
+/// - All whitespace will be trimmed: ` abc/def ` => `abc/def`
```

Comment:
Here the doc contradict with the actual behavior, ` abc/def ` will be normalized to `abc/def/` as per the implementation.

Reply:
It is just one of the rules; the function will apply all rules and yield the final result.